File: packages/rust/src/tokenize.rs
// SPDX-License-Identifier: MIT
//! Pure-Rust BPE encoder. Text → token IDs.
//!
//! Required for the bidirectional Codec endpoint where the client wants
//! to send token-ID prompts (zero text on the wire in either direction).
//!
//! ## Algorithm (for both byte_level and metaspace BPE)
//!
//! 1. Pre-tokenize: split input into pieces (regex for byte_level; whitespace for metaspace).
//! 2. Encode each piece into the vocab's character space (GPT-2 byte chars or `▁`-prefixed).
//! 3. Apply BPE merges greedily by priority — match HuggingFace reference.
//! 4. Look up final tokens in `vocab`. Tokens not in vocab fall back to byte tokens (metaspace path).

use std::cell::RefCell;
use std::collections::HashMap;

use regex::Regex;

use crate::byte_encoder::{encode_byte_level_chars, METASPACE};
use crate::map::TokenizerMap;

/// Common interface every tokenizer implementation satisfies.
///
/// Implemented by [`BPETokenizer`] and [`crate::longest_match::LongestMatchTokenizer`].
///
/// The trait deliberately does not require `Sync` — `BPETokenizer` keeps
/// a `RefCell`-backed encode cache (mirroring the .NET `Dictionary`).
/// Wrap in `Mutex` for cross-thread sharing.
pub trait ITokenizer: Send {
    /// Identifier of the underlying vocabulary.
    fn id(&self) -> &str;
    /// Encode a string to a sequence of token IDs.
    fn encode(&self, text: &str) -> Vec<u32>;
}

/// Pure-Rust BPE encoder.
///
/// Construct via [`BPETokenizer::new`]; check
/// [`BPETokenizer::supports`] first if you don't know whether the map has
/// the data BPE needs (use [`crate::Tokenize::pick`] which falls back).
pub struct BPETokenizer {
    id: String,
    vocab: HashMap<String, u32>,
    /// Map of `"left right"` → priority rank (lower = higher priority).
    merge_ranks: HashMap<String, u32>,
    pre_tok_regex: Option<Regex>,
    encoder: String,
    /// `i64` so a missing fallback (-1) is comparable safely against IDs.
    byte_fallback_start: i64,
    /// Per-piece encode cache. Mutex-free RefCell — BPETokenizer is `!Sync`
    /// but `Send`. (Translator only needs `Send` and constructs its own.)
    cache: RefCell<HashMap<String, Vec<u32>>>,
}

impl BPETokenizer {
    /// Returns true if the map has the data BPETokenizer needs.
    pub fn supports(map: &TokenizerMap) -> bool {
        let has_vocab = map.vocab.as_ref().is_some_and(|v| !v.is_empty());
        let has_merges = map.merges.as_ref().is_some_and(|v| !v.is_empty());
        let enc_ok = matches!(map.encoder.as_deref(), Some("byte_level") | Some("metaspace"));
        has_vocab && has_merges && enc_ok
    }

    /// Construct a `BPETokenizer` from a map. Returns an error message if
    /// the map lacks the required vocab/merges/encoder.
    pub fn new(map: &TokenizerMap) -> Result<Self, String> {
        if !Self::supports(map) {
            return Err(format!(
                "BPETokenizer: map \"{}\" lacks vocab/merges/encoder. \
                 Use BPETokenizer::supports(map) to check first, or call \
                 Tokenize::pick(map) which falls back to LongestMatchTokenizer.",
                map.id
            ));
        }

        let vocab = map.vocab.as_ref().expect("supports() checked").clone();
        let merges = map.merges.as_ref().expect("supports() checked");
        let encoder = map.encoder.as_ref().expect("supports() checked").clone();
        let id = map.id.clone();
        let byte_fallback_start = map.byte_fallback_start.unwrap_or(-1);

        let mut merge_ranks: HashMap<String, u32> = HashMap::with_capacity(merges.len());
        for (i, m) in merges.iter().enumerate() {
            merge_ranks.insert(m.clone(), i as u32);
        }

        let pre_tok_regex = if encoder == "byte_level" {
            let pat = map.pre_tokenizer_pattern.as_ref().ok_or_else(|| {
                format!(
                    "BPETokenizer: byte_level map \"{}\" missing pre_tokenizer_pattern.",
                    map.id
                )
            })?;
            Some(
                Regex::new(pat)
                    .map_err(|e| format!("BPETokenizer: invalid pre_tokenizer_pattern: {e}"))?,
            )
        } else {
            None
        };

        Ok(Self {
            id,
            vocab,
            merge_ranks,
            pre_tok_regex,
            encoder,
            byte_fallback_start,
            cache: RefCell::new(HashMap::new()),
        })
    }

    /// Encode text → token IDs.
    pub fn encode(&self, text: &str) -> Vec<u32> {
        if text.is_empty() {
            return Vec::new();
        }

        let pieces = self.pre_tokenize(text);
        let mut ids: Vec<u32> = Vec::with_capacity(pieces.len() * 2);
        for piece in pieces {
            // Cache hit?
            if let Ok(cache) = self.cache.try_borrow() {
                if let Some(cached) = cache.get(&piece) {
                    ids.extend_from_slice(cached);
                    continue;
                }
            }
            let encoded = self.encode_piece_to_vocab_space(&piece);
            let merged = self.apply_bpe(encoded);
            let piece_ids = self.lookup(&merged);
            if let Ok(mut cache) = self.cache.try_borrow_mut() {
                cache.insert(piece.clone(), piece_ids.clone());
            }
            ids.extend_from_slice(&piece_ids);
        }
        ids
    }

    // ── Pre-tokenization ────────────────────────────────────────────────────

    fn pre_tokenize(&self, text: &str) -> Vec<String> {
        if self.encoder == "byte_level" {
            let re = self.pre_tok_regex.as_ref().expect("byte_level requires regex");
            return re.find_iter(text).map(|m| m.as_str().to_string()).collect();
        }

        // Metaspace: split on whitespace, prefix every word with ▁.
        // Collapse internal runs of spaces/tabs to a single space first
        // (matches .NET).
        let collapsed = collapse_spaces_and_tabs(text);
        let parts = split_keep_whitespace(&collapsed);
        let mut pieces: Vec<String> = Vec::new();
        for p in parts {
            if p == " " {
                continue;
            }
            let mut s = String::with_capacity(p.len() + 3);
            s.push(METASPACE);
            s.push_str(&p);
            pieces.push(s);
        }
        pieces
    }

    // ── Step 2: piece → vocab character space ──────────────────────────────

    fn encode_piece_to_vocab_space(&self, piece: &str) -> Vec<String> {
        if self.encoder == "byte_level" {
            let bytes = piece.as_bytes();
            let encoded = encode_byte_level_chars(bytes);
            return codepoints(&encoded);
        }
        codepoints(piece)
    }

    // ── Step 3: BPE merges ─────────────────────────────────────────────────

    fn apply_bpe(&self, tokens: Vec<String>) -> Vec<String> {
        if tokens.len() < 2 {
            return tokens;
        }
        let mut parts = tokens;
        loop {
            let mut best_idx: Option<usize> = None;
            let mut best_rank: u32 = u32::MAX;
            for i in 0..parts.len() - 1 {
                // Build "left right" without alloc churn — small strings only here.
                let mut key = String::with_capacity(parts[i].len() + 1 + parts[i + 1].len());
                key.push_str(&parts[i]);
                key.push(' ');
                key.push_str(&parts[i + 1]);
                if let Some(&r) = self.merge_ranks.get(&key) {
                    if r < best_rank {
                        best_rank = r;
                        best_idx = Some(i);
                    }
                }
            }
            let Some(_idx) = best_idx else {
                break;
            };

            let left = parts[best_idx.unwrap()].clone();
            let right = parts[best_idx.unwrap() + 1].clone();
            let merged = format!("{left}{right}");

            // Merge ALL non-overlapping occurrences in one pass — matches HF.
            let mut next: Vec<String> = Vec::with_capacity(parts.len());
            let mut j = 0;
            while j < parts.len() {
                if j + 1 < parts.len() && parts[j] == left && parts[j + 1] == right {
                    next.push(merged.clone());
                    j += 2;
                } else {
                    next.push(parts[j].clone());
                    j += 1;
                }
            }
            parts = next;
        }
        parts
    }

    // ── Step 4: vocab lookup with byte fallback ────────────────────────────

    fn lookup(&self, tokens: &[String]) -> Vec<u32> {
        let mut ids: Vec<u32> = Vec::with_capacity(tokens.len());
        for tok in tokens {
            if let Some(&id) = self.vocab.get(tok) {
                ids.push(id);
                continue;
            }
            if self.byte_fallback_start >= 0 {
                for &b in tok.as_bytes() {
                    ids.push((self.byte_fallback_start + b as i64) as u32);
                }
            }
            // For byte_level this is unreachable for valid UTF-8 input.
        }
        ids
    }
}

impl ITokenizer for BPETokenizer {
    fn id(&self) -> &str {
        &self.id
    }
    fn encode(&self, text: &str) -> Vec<u32> {
        BPETokenizer::encode(self, text)
    }
}

// Helpers ------------------------------------------------------------------

fn codepoints(s: &str) -> Vec<String> {
    s.chars().map(|c| c.to_string()).collect()
}

fn collapse_spaces_and_tabs(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut prev_space = false;
    for c in s.chars() {
        if c == ' ' || c == '\t' {
            if !prev_space {
                out.push(' ');
                prev_space = true;
            }
        } else {
            out.push(c);
            prev_space = false;
        }
    }
    out
}

/// Split on whitespace, keeping each whitespace char as its own segment
/// (mirrors .NET `Regex.Split(text, "(\\s)")`).
fn split_keep_whitespace(s: &str) -> Vec<String> {
    let mut parts: Vec<String> = Vec::new();
    let mut buf = String::new();
    for c in s.chars() {
        if c.is_whitespace() {
            if !buf.is_empty() {
                parts.push(std::mem::take(&mut buf));
            }
            parts.push(c.to_string());
        } else {
            buf.push(c);
        }
    }
    if !buf.is_empty() {
        parts.push(buf);
    }
    parts
}

File: packages/python/src/codecai/tokenize.py
"""Text → token IDs.

Two tokenizer implementations:

- :class:`BPETokenizer`: pure-Python BPE, byte_level + metaspace. Use for
  any v2 map that has merges (every map fetched from ``codec-maps`` for a
  real model).
- :class:`LongestMatchTokenizer`: vocab-only longest-prefix-match. Suitable
  for canonical-IR / synthetic test maps.

:func:`pick_tokenizer` returns the right one for a given map.
"""
from __future__ import annotations

import re as _stdlib_re
from abc import ABC, abstractmethod
from typing import Any

# Use the third-party `regex` module for the byte_level pre-tokenizer pattern,
# which uses Unicode property classes (\p{L}, \p{N}) that stdlib `re` doesn't
# support. Falls back to `re` for the simpler whitespace-collapsing patterns.
import regex

from .encoder import METASPACE, encode_byte_level_chars
from .pretok_program import run_pretok_program
from .types import TokenizerMap

_DELIMITER_BODY = _stdlib_re.compile(r"^[A-Za-z0-9_-]+$")


def _is_delimiter_shape(tok: str) -> bool:
    """Match ``<|body|>`` where body is non-empty and identifier-like.

    Catches every shipped chat-template and tool-call delimiter while
    excluding pathological vocab BPE tokens like Falcon's ``<|>`` (id
    61799) that share the start/end pair.
    """
    if len(tok) <= 4:
        return False
    if not (tok.startswith("<|") and tok.endswith("|>")):
        return False
    return bool(_DELIMITER_BODY.match(tok[2:-2]))


class Tokenizer(ABC):
    """Common interface every tokenizer satisfies."""

    @property
    @abstractmethod
    def id(self) -> str:
        ...

    @abstractmethod
    def encode(self, text: str) -> list[int]:
        ...


# ── BPE ──────────────────────────────────────────────────────────────────────


class BPETokenizer(Tokenizer):
    """Pure-Python BPE encoder. Text → token IDs.

    Algorithm (for both byte_level and metaspace BPE):

    1. Pre-tokenize: split input text into pieces (regex for byte_level,
       whitespace for metaspace).
    2. Encode each piece into the vocab's character space (GPT-2 byte chars
       or ``▁``-prefixed).
    3. Apply BPE merges greedily by priority: match HuggingFace reference.
    4. Look up final tokens in :attr:`vocab`. Tokens not in the vocab fall
       back to byte tokens (metaspace path).
    """

    __slots__ = (
        "_id", "_encoder", "_vocab", "_merge_ranks", "_pre_tok_re",
        "_pre_tok_program",
        "_byte_fallback_start", "_cache",
        "_special_ids", "_special_re",
    )

    @staticmethod
    def supports(m: TokenizerMap) -> bool:
        """True if the map has the data this tokenizer needs."""
        return bool(
            m.vocab
            and m.merges
            and m.encoder in ("byte_level", "metaspace")
        )

    def __init__(self, m: TokenizerMap) -> None:
        if not BPETokenizer.supports(m):
            raise ValueError(
                f"BPETokenizer: map {m.id!r} lacks vocab/merges/encoder. "
                "Use BPETokenizer.supports(map) to check first, or call "
                "pick_tokenizer(map) which falls back to LongestMatchTokenizer."
            )

        assert m.vocab is not None
        assert m.merges is not None

        self._id = m.id
        self._encoder = m.encoder
        self._vocab: dict[str, int] = dict(m.vocab)
        self._merge_ranks: dict[str, int] = {entry: i for i, entry in enumerate(m.merges)}
        self._byte_fallback_start = m.byte_fallback_start if m.byte_fallback_start is not None else -1
        self._cache: dict[str, list[int]] = {}

        # Pre-tokenizer: prefer the compiled program when present, otherwise
        # fall back to the legacy regex. Programs are required for clients
        # without a Unicode regex engine (libcodec/C); Python already has
        # one via the `regex` package, so the program here is mostly a
        # startup-time speedup plus keeping every client on the same
        # code path, which is what makes the equivalence claim auditable.
        if self._encoder == "byte_level":
            if m.pre_tokenizer_program and m.pre_tokenizer_program.get("ops"):
                self._pre_tok_program: dict[str, Any] | None = m.pre_tokenizer_program
                self._pre_tok_re = None
            elif m.pre_tokenizer_pattern:
                self._pre_tok_re = regex.compile(m.pre_tokenizer_pattern)
                self._pre_tok_program = None
            else:
                raise ValueError(
                    f"BPETokenizer: byte_level map {m.id!r} missing both "
                    "pre_tokenizer_program and pre_tokenizer_pattern."
                )
        else:
            self._pre_tok_re = None
            self._pre_tok_program = None

        # Build the special-token scanner. Accept entries from
        # ``special_tokens`` AND any vocab key in ``<|body|>`` shape where
        # body is non-empty and identifier-like: older maps shipped
        # before a chat-template revision may carry the delimiters in
        # vocab but not in special_tokens. Length-descending order makes
        # the regex match the longest delimiter at any position. Without
        # this pre-scan, ``<|im_start|>`` would tokenise byte-by-byte
        # instead of as the single atomic vocab ID. The body constraint
        # excludes pathological vocab tokens like Falcon's ``<|>`` (id
        # 61799) that share the start/end pair.
        special: dict[str, int] = dict(m.special_tokens or {})
        for tok, tid in self._vocab.items():
            if tok in special:
                continue
            if _is_delimiter_shape(tok):
                special[tok] = tid
        self._special_ids: dict[str, int] = special
        if special:
            keys = sorted(special.keys(), key=len, reverse=True)
            self._special_re: regex.Pattern[str] | None = regex.compile(
                "|".join(regex.escape(k) for k in keys)
            )
        else:
            self._special_re = None

    @property
    def id(self) -> str:
        return self._id

    def encode(self, text: str) -> list[int]:
        if not text:
            return []

        if self._special_re is not None:
            out: list[int] = []
            cursor = 0
            for m in self._special_re.finditer(text):
                if m.start() > cursor:
                    self._encode_chunk(text[cursor:m.start()], out)
                out.append(self._special_ids[m.group(0)])
                cursor = m.end()
            if cursor < len(text):
                self._encode_chunk(text[cursor:], out)
            return out

        ids: list[int] = []
        self._encode_chunk(text, ids)
        return ids

    def _encode_chunk(self, text: str, out: list[int]) -> None:
        if not text:
            return
        for piece in self._pre_tokenize(text):
            cached = self._cache.get(piece)
            if cached is not None:
                out.extend(cached)
                continue
            encoded = self._encode_piece_to_vocab_space(piece)
            merged = self._apply_bpe(encoded)
            piece_ids = self._lookup(merged)
            self._cache[piece] = piece_ids
            out.extend(piece_ids)

    # ── Pre-tokenization ────────────────────────────────────────────────────

    def _pre_tokenize(self, text: str) -> list[str]:
        if self._encoder == "byte_level":
            if self._pre_tok_program is not None:
                return run_pretok_program(self._pre_tok_program, text)
            assert self._pre_tok_re is not None
            return [m for m in self._pre_tok_re.findall(text) if m]

        # Metaspace: split on whitespace, prefix every word with ▁.
        # Mirrors the JS implementation: collapse runs of spaces/tabs to
        # single spaces, then attach ▁ to each non-whitespace piece.
        collapsed = _stdlib_re.sub(r"[ \t]+", " ", text)
        out: list[str] = []
        for part in _stdlib_re.split(r"(\s)", collapsed):
            if not part or part == " ":
                continue
            out.append(METASPACE + part)
        return out

    # ── Step 2: piece → vocab character space ──────────────────────────────

    def _encode_piece_to_vocab_space(self, piece: str) -> list[str]:
        if self._encoder == "byte_level":
            return list(encode_byte_level_chars(piece.encode("utf-8")))
        # metaspace: piece is already in vocab space
        return list(piece)

    # ── Step 3: BPE merges ─────────────────────────────────────────────────

    def _apply_bpe(self, tokens: list[str]) -> list[str]:
        if len(tokens) < 2:
            return tokens

        parts = list(tokens)
        while True:
            best_idx = -1
            best_rank = float("inf")
            for i in range(len(parts) - 1):
                key = parts[i] + " " + parts[i + 1]
                rank = self._merge_ranks.get(key)
                if rank is not None and rank < best_rank:
                    best_rank = rank
                    best_idx = i
            if best_idx == -1:
                break

            # Merge ALL non-overlapping occurrences of that pair in one pass:
            # matches HuggingFace.
            left = parts[best_idx]
            right = parts[best_idx + 1]
            merged = left + right
            new_parts: list[str] = []
            i = 0
            while i < len(parts):
                if i < len(parts) - 1 and parts[i] == left and parts[i + 1] == right:
                    new_parts.append(merged)
                    i += 2
                else:
                    new_parts.append(parts[i])
                    i += 1
            parts = new_parts

        return parts

    # ── Step 4: vocab lookup with byte fallback ────────────────────────────

    def _lookup(self, tokens: list[str]) -> list[int]:
        ids: list[int] = []
        for tok in tokens:
            tok_id = self._vocab.get(tok)
            if tok_id is not None:
                ids.append(tok_id)
                continue
            if self._byte_fallback_start >= 0:
                # Metaspace + SentencePiece byte_fallback: emit raw UTF-8 bytes.
                for b in tok.encode("utf-8"):
                    ids.append(self._byte_fallback_start + b)
            # For byte_level this branch is unreachable for valid UTF-8 input.
        return ids


# ── LongestMatch ─────────────────────────────────────────────────────────────


class LongestMatchTokenizer(Tokenizer):
    """Vocab-only longest-prefix-match tokenizer. Fallback for non-BPE maps."""

    __slots__ = ("_id", "_fragment_to_id", "_max_fragment_length", "_special_fragment_to_id")

    def __init__(self, m: TokenizerMap) -> None:
        self._id = m.id
        self._fragment_to_id: dict[str, int] = {}
        max_len = 1

        if m.vocab is not None:
            for fragment, tok_id in m.vocab.items():
                if not fragment:
                    continue
                self._fragment_to_id[fragment] = tok_id
                if len(fragment) > max_len:
                    max_len = len(fragment)
        if m.tokens is not None:
            for id_str, fragment in m.tokens.items():
                if not fragment:
                    continue
                try:
                    tok_id = int(id_str)
                except ValueError:
                    continue
                self._fragment_to_id[fragment] = tok_id
                if len(fragment) > max_len:
                    max_len = len(fragment)

        self._max_fragment_length = max_len
        self._special_fragment_to_id: dict[str, int] = {}
        if m.special_tokens is not None:
            for name, tok_id in m.special_tokens.items():
                self._special_fragment_to_id[name] = tok_id
                if not name.startswith("<"):
                    self._special_fragment_to_id[f"<|{name}|>"] = tok_id

    @property
    def id(self) -> str:
        return self._id

    def encode(self, text: str) -> list[int]:
        out: list[int] = []
        pos = 0
        n = len(text)
        while pos < n:
            consumed = False
            for frag, tok_id in self._special_fragment_to_id.items():
                if text.startswith(frag, pos):
                    out.append(tok_id)
                    pos += len(frag)
                    consumed = True
                    break
            if consumed:
                continue

            remaining = n - pos
            try_up_to = min(self._max_fragment_length, remaining)
            matched_id = -1
            matched_len = 0
            for length in range(try_up_to, 0, -1):
                candidate = text[pos:pos + length]
                tok_id = self._fragment_to_id.get(candidate)
                if tok_id is not None:
                    matched_id = tok_id
                    matched_len = length
                    break

            if matched_id == -1:
                out.append(0)  # UNK by convention
                pos += 1
            else:
                out.append(matched_id)
                pos += matched_len
        return out


# ── Factory ──────────────────────────────────────────────────────────────────


def pick_tokenizer(m: TokenizerMap) -> Tokenizer:
    """Build the right tokenizer for the map."""
    if BPETokenizer.supports(m):
        return BPETokenizer(m)
    return LongestMatchTokenizer(m)


def tokenize(m: TokenizerMap, text: str) -> list[int]:
    """One-shot encode using :func:`pick_tokenizer`."""
    return pick_tokenizer(m).encode(text)
